common-ui-compose/src/main/java/app/tivi/common/compose/insets.kt

```diff
@@ -14,64 +14,410 @@
-    val vertical get() = top + bottom
+/**
+ * Main holder of our inset values.
+ */
+@Stable
+class DisplayInsets {
+    val systemBars = Insets()
+    val systemGestures = Insets()
 }
 
-val InsetsAmbient = ambientOf { InsetsHolder() }
+@Stable
+class Insets {
+    var left by mutableStateOf(0)
+        internal set
+    var top by mutableStateOf(0)
+        internal set
+    var right by mutableStateOf(0)
+        internal set
+    var bottom by mutableStateOf(0)
+        internal set
 
+    /**
+     * TODO: doesn't currently work
+     */
+    var visible by mutableStateOf(true)
+        internal set
+}
+
+val DisplayInsetsAmbient = staticAmbientOf<DisplayInsets>()
+
+/**
+ * Setup and provide [DisplayInsets] to [DisplayInsetsAmbient].
+ *
+ * This will automatically listen for [WindowInsetsCompat] on the host view.
+ *
+ * @param consumeInsets Whether to consume the insets which are dispatched to the host view.
+ */
 @Composable
-fun ProvideInsets(
-    liveData: LiveData<WindowInsetsCompat?>,
-    children: @Composable () -> Unit
+fun ProvideDisplayInsets(
+    consumeInsets: Boolean = true,
+    content: @Composable () -> Unit
 ) {
-    val currentInsets = remember(liveData) {
-        liveData.map { if (it != null) InsetsHolder(it) else InsetsHolder() }
-            .distinctUntilChanged()
-    }.observeAsState(InsetsHolder())
+    val view = ViewAmbient.current
+
+    val displayInsets = remember { DisplayInsets() }
+
+    onCommit(view) {
+        ViewCompat.setOnApplyWindowInsetsListener(view) { _, windowInsets ->
+            displayInsets.systemBars.updateFrom(windowInsets.systemWindowInsets)
+            displayInsets.systemGestures.updateFrom(windowInsets.systemGestureInsets)
+
+            if (consumeInsets) {
+                // TODO: Move this to WindowInsetsCompat.CONSUMED
+                windowInsets.consumeSystemWindowInsets()
+            } else {
+                // Return the unconsumed insets
+                windowInsets
+            }
+        }
+
+        // Add an OnAttachStateChangeListener to request an inset pass each time we're attached
+        // to the window
+        val attachListener = object : View.OnAttachStateChangeListener {
+            override fun onViewAttachedToWindow(v: View) = v.requestApplyInsets()
+            override fun onViewDetachedFromWindow(v: View) = Unit
+        }
+        view.addOnAttachStateChangeListener(attachListener)
+
+        if (view.isAttachedToWindow) {
+            // If the view is already attached, we can request an inset pass now
+            view.requestApplyInsets()
+        }
+
+        onDispose {
+            // Once we leave the composition, remove our OnAttachStateChangeListener
+            view.removeOnAttachStateChangeListener(attachListener)
+        }
+    }
+
+    Providers(DisplayInsetsAmbient provides displayInsets) {
+        content()
+    }
+}
+
+/**
+ * Selectively apply additional space which matches the width/height of any system bars present
```

Comment:
What about replacing 'additional space' with 'padding' throughout. Right now comments mix these terms. With `Modifier.padding` being the lib's chosen naming, seems to make sense to move to padding?

Reply:
Done.